zrml/liquidity-mining/src/distribute_incentives_based_on_bought_shares.rs

```diff
@@ -0,0 +1,108 @@
+                let raw_incentives = share_value.saturating_mul(bought_shares);
+                let perpetual_incentives = calculate_perthousand_value(ppb, raw_incentives);
+                let incentives = raw_incentives.saturating_sub(perpetual_incentives);
+                <OwnedValues<T>>::mutate(market_id, account_id, |values| {
+                    let one = T::BlockNumber::from(1u8);
+
+                    values.perpetual_incentives = values
+                        .perpetual_incentives
+                        .saturating_add(perpetual_incentives);
+                    values.total_incentives = values.total_incentives.saturating_add(incentives);
+                    values.total_shares = values.total_shares.saturating_add(bought_shares);
+                    values.participated_blocks = values.participated_blocks.saturating_add(one);
+                });
+            }
+        }
+
+        <BlockBoughtShares<T>>::remove_all();
+    }
+
+    // ZTG value of one bought share for the current block being produced. Or in other words:
+    // Determines how much a share will be worth given the amount of ZTG for liquidity
+    // mining and the total number of bought shares for the current block.
+    //
+    // `None` result means no-one purchased a share.
+    #[inline]
+    fn bought_share_value(
+        incentive: &BalanceOf<T>,
+        total_bought_shares: &BalanceOf<T>,
+    ) -> Option<BalanceOf<T>> {
+        incentive.checked_div(&total_bought_shares)
+    }
+
+    // How much incentive each market will receive
+    fn markets_incentives(
+        per_block_incentives: BalanceOf<T>,
+        curr_block: T::BlockNumber,
+    ) -> Vec<(T::MarketId, BalanceOf<T>)> {
+        let mut normalized_total = BalanceOf::<T>::from(0u8);
+        let normalized_values: Vec<_> = <Markets<T>>::iter()
+            .map(|(market_id, period)| {
+                let normalized_value = Self::normalize_market(curr_block, period);
+                normalized_total = normalized_total.saturating_add(normalized_value);
+                (market_id, normalized_value)
+            })
+            .collect();
+        normalized_values
+            .into_iter()
+            .filter_map(|(market_id, normalized_value)| {
+                let incentive_ptd = calculate_perthousand(normalized_value, &normalized_total)?;
+                let incentive = calculate_perthousand_value(incentive_ptd, per_block_incentives);
+                Some((market_id, incentive))
+            })
+            .collect()
+    }
+
+    // Takes any market parameter and outputs a number that will be used as a percentage
+    // to calculate how much incentives each individual market will receive.
+    //
+    // In this case, the output is the percentage of the remaining number of blocks to stimulate
+    // early liquidity providers. For example, a market starts at 0 and ends at 10. If the current
+    // block is at 7, then the output is 30.
```

Comment:
Can you give some additional examples of this? Not sure I totally understand how this will differ the payout to LPs that enter at different blocks.

Reply:
Added one more example. Please, feel free to discuss any doubts about the algorithm